Add static file upload

zwade opened the pull request:
[Resolves #105] Includes a way to upload from the editor, and a standalone admin file manager.

zwade committed:
got simple file uploading working

zwade committed:
Added a files page for managing uploaded files

zwade committed:
added all files this time

bluepichu commented on client/hbs/editor.hbs:
```diff
@@ -34,9 +34,18 @@
 			</footer>
 		</main>
+		<div class='uploader'>
+			<button id='uploadTrigger' class='upload-button upload-item'>Upload New File</button>
```
The `id` here doesn't follow the convention used elsewhere.  I recommend switching it to `upload-trigger` for consistency.

bluepichu commented on client/hbs/editor.hbs:
```diff
@@ -34,9 +34,18 @@
+			</a>
+		</div>
+		<form action='/static/' method='POST' class='hidden'>
+			<input type='file' name='file' id='fileUpload'>
```
See above; `s/fileUpload/file-upload`

bluepichu commented on client/hbs/files.hbs:
```diff
@@ -0,0 +1,31 @@
+{{#if user}}
+	<header>
+		<h1>File Browser</h1>
+		<a id="filePageTrigger">
```
`s/filePageTrigger/file-page-trigger`

bluepichu commented on client/hbs/editor.hbs:
```diff
@@ -34,9 +34,18 @@
+			<button id='uploadTrigger' class='upload-button upload-item'>Upload New File</button>
+			<div class='upload-name upload-item disabled'></div>
+			<a href='/files/' target='_blank' class='button upload-launch upload-item'>
+				Browse <i class="material-icons">launch</i>
```
`<i>` is not the semantically correct choice here.

Also, `<dttw-icon>` should work? :)

bluepichu commented on client/hbs/files.hbs:
```diff
@@ -0,0 +1,31 @@
+	<header>
+		<h1>File Browser</h1>
+		<a id="filePageTrigger">
+			<dttw-icon id="upload" class="material-icons">file_upload</dttw-icon>
```
`<dttw-icon>` shouldn't need `class="material-icons"`, iirc.

bluepichu replied:
Actually, looks like I'm wrong, but I don't think that there's a case where `<dttw-icon>` shouldn't be `.material-icons`, so I'm not sure why it's set up this way.  We'll handle this later, I guess?

bluepichu commented on client/js/admin.js:
```diff
@@ -38,7 +39,86 @@ $(function() {
+			XHR.open("POST", "/static/")
+			var fData = new FormData(form[0].form)
+			XHR.onload = function() {
+				var res = $(".upload-name.upload-item")
```
Any reason to not simply `$(".upload-name.upload-item").removeClass("disabled").text(out.path)`?

zwade replied:
huh, didn't realise you could do that

bluepichu commented on client/js/admin.js:
```diff
@@ -38,7 +39,86 @@ $(function() {
+			}
+			XHR.send(fData)
+		})
+		form.click()
```
This feels wrong.  It might _work_, but that doesn't sit right with me.

Maybe `form.submit()`?

zwade replied:
oh no, this is actually the proper way to do it :P

bluepichu commented on client/js/admin.js:
```diff
@@ -38,7 +39,86 @@ $(function() {
+			XHR.open("POST", "/static/")
+			var fData = new FormData(form[0].form)
+			XHR.onload = function() {
+				var dummy = $(".file-item:first-child")[0].outerHTML
```
I don't know what's happening here, but I definitely don't like it.  Replaces on raw HTML are risky and generally unneeded.

zwade replied:
Oh, this element is a dummy placed in there. Its a way to copy the server side template without so I can do a client side animation. Only necessary if you want to reload the page after file upload. 

Bear in mind, this is close to how templating engines work.

bluepichu replied:
How about this instead:

```
// remove the dummy element before doing this
let fileItem = $(".file-item:first-child").clone();
fileItem.find("a:first-child").attr("href", out.path);
fileItem.find("a.copy-path").attr("path", out.path);
fileItem.find("span.title").text(out.file);
fileItem.find("a.delete-path").attr("path", out.delete);
```

bluepichu commented on client/scss/admin.scss:
```diff
@@ -61,20 +62,23 @@ header {
 		margin: 0px .4em;
+	}
+}
+a > dttw-icon {
```
Could this selector be more specific?

zwade replied:
Potentially, what don't you like about this?

bluepichu replied:
Never mind, I'm fine with it.  I misunderstood what was going on.

bluepichu commented on client/hbs/editor.hbs:
```diff
@@ -34,9 +34,18 @@
+		<div class='uploader'>
+			<button id='uploadTrigger' class='upload-button upload-item'>Upload New File</button>
+			<div class='upload-name upload-item disabled'></div>
+			<a href='/files/' target='_blank' class='button upload-launch upload-item'>
```
Why is `.button` a thing?

zwade replied:
so I can use href on a button thing

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+		uploadPost(null, req.body, req.user.id)
+			.then(function (id) {
+				renderer.reload()
+				res.send({msg: "Ok", id: id})
```
What's the point of the useless `msg` field?

zwade replied:
lol this isn't actually part of the diff, I just in a `requireAdmin` statement for simplicity

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+	  * Annotations
+	**/ 
+
+	var requireAdmin = (fn) =>
```
Why is this not just `function requireAdmin`?  I don't see a scoping concern here.

zwade replied:
I prefer to use arrow functions for curried functions, mostly because it's more clear in my view

bluepichu replied:
Ok... on the flip-side, why not `(fn) => (req, res) => ...`?

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+	app.delete("/static/:FILE", requireAdmin(function(req, res) {
+		var filePath = path.join(__dirname, config.paths.upload, req.params.FILE)
+		fs.stat(filePath).then( (stat) => {
+			if (stat || stat.isFile()) {
```
Should this not be `stat && stat.isFile`?

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+				//All Good!
+				return
+			} else {
+				throw new Error("File Not Found")
```
I think a 404 response is the correct action here, not a 500.

bluepichu replied:
Actually, you're not sending any response at all, if I'm reading this correctly.  It'll just spin.

zwade replied:
oops ._.

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+				throw new Error("File Not Found")
+			}
+		}).then(fs.unlink(filePath)).then( () => {
+			return res.send({success: true})
```
If `success` is never `false`, why is it here?

zwade replied:
That was a result of me forgetting to add in the failure condition XD

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+			})
+	}))
 
+	app.post("/static/", upload.single('file'), requireAdmin(function(req, res) {
```
Should the path not just be `/static`?

bluepichu replied:
Also, I'm confused why the uploading endpoint is called `static`.

zwade replied:
Idk, seemed reasonable since you were modifying a static file. I can change it if you feel strongly

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+	}))
 
+	app.post("/static/", upload.single('file'), requireAdmin(function(req, res) {
+		let hash = req.file.filename.slice(0,8)
```
...am I missing something here?  How is this a hash?

zwade replied:
Yeah, the library performs a hash and sets that as the filename. Sllly naming, but my code is accurate

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+					.split(".")
+					.map((e) => utils.slugify(e))
+					.join(".")
+		/** No directory traversals in my house **/
```
Please replace with line comment.

`// No unnecessary block comments in my house`

zwade replied:
but block comments are pretttttier

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+					.map((e) => utils.slugify(e))
+					.join(".")
+		/** No directory traversals in my house **/
+		sluggedFile = sluggedFile.replace(/\.(\.)+/g, "")
```
Couldn't you skip this step by doing `.join("")` above?  That's the only way to reintroduce `.` into the string.

zwade replied:
lol but then you wouldn't have a file extension XD

bluepichu commented on server/server.js:
```diff
@@ -51,64 +57,95 @@ module.exports = function(__dirname) {
+			.then( () => res.send({path: path.join("/upload/", prefixName), 
+			                       file: prefixName,
+				               shortFile: sluggedFile}) )
+			.catch( (e) => {
```
Your use of spacing here confuses and upsets me.

Technically not wrong but wut

zwade replied:
haha sorry, that's the python spec. Apparently I got confused.

bluepichu replied:
You "fixed" the wrong thing.  I like the hanging `.catch`, it's more readable.

I was more confused by the space between `.catch(` and `(e) => {`.

bluepichu commented on client/scss/admin.scss:
```diff
@@ -346,4 +435,8 @@ input[type="checkbox"] {
+}
+
+.hidden {
+	opacity: 0;
```
Justification of this over `display: none`, if any?

zwade replied:
Copying won't work if the text node is hidden

el1t replied:
ambiguous class name with `visibility: hidden` tho

zwade replied:
yeah, that was I originally did, until I realized it didn't work

zwade committed:
Semantic Fixes

zwade committed:
addressing semantic issues

zwade committed:
more fixes for @bluepichu

zwade committed:
fixed bugs with upload

bluepichu commented:
Hidden post markers are broken, probably due to an unspecific selector on `.hidden`:

![screenshot 2016-07-17 11 15 21](https://cloud.githubusercontent.com/assets/1878074/16901373/c8978d88-4c0f-11e6-9426-ced9731536fb.png)

This is why I think we need end-to-end tests :)

bluepichu commented:
The file uploader doesn't make sense to me.  You click "upload file", then choose your file... then the file name gets filled in on the line for some reason?  And there's no confirmation?

bluepichu commented:
![screenshot 2016-07-17 11 22 38](https://cloud.githubusercontent.com/assets/1878074/16901406/cd96541c-4c10-11e6-924b-6d918ba1b67c.png)

The left button doesn't do what I'd expect (i.e. download the file without showing it to me), and the right button doesn't do anything at all.

bluepichu commented on client/hbs/editor.hbs:
```diff
@@ -34,9 +34,20 @@
 				</button>
 			</footer>
 		</main>
+		<div class='uploader'>
```
afaict `.upload-item` is a useless selector.  Everything inside `.uploader` (which we might want to be `#uploader`?) is a `.upload-item`, and nothing else is.

zwade replied:
oops

zwade replied:
(fixed in 7eaa94f)

bluepichu commented:
Uploading a file from the file viewer results in an inconsistent ordering: the file is added at the bottom, but on refresh, it appears at the top.

bluepichu commented on server/routes.js:
```diff
@@ -259,4 +271,22 @@ function fillDefaultSidebar(db) {
+				delete: path.join("/static/", f), 
+				name: f.split("-").splice(1).join("-")
+			}})
+			context.files.unshift({
```
I just realized what is going on here.  Please no lying about what files we have.

(See later comment on the pseudo-templating)

zwade replied:
So I'm not sure how you would prefer me to do it, assuming that I don't think it would be good to
- Copy the templating source
- Force a reload on upload
- Do 100% client side templating

This seemed the best solution to me

zwade commented:
Addressing your comments in reverse order

> Uploading a file from the file browser results in an inconsistent ordering: the file is added at the bottom, but on refresh, it appears at the top.

The new element always gets put at the bottom, since its not defined as to which position it would go in, since it pulls the array from `fs.dirstat`. Now, I could sort the resulting array before building the template, but that didn't seem like a big deal to me

> Also, I'm not sure if the upload timestamp is part of the hash, but sometimes I can upload duplicates and sometimes I can't.

It should get renamed if you try to upload a dupe. When is it not working for you?

> Also, the hash is now missing from the filename, which is weird.

I explicitly removed the hash to make it easier to see which file is which. I prefer it that way, but I can add the hash if you want. 

> The left button doesn't do what I'd expect (i.e. download the file without showing it to me), and the right button doesn't seem to do anything at all.

The left button serves the file with its proper mime type. The reason for this is so you can serve it directly. I'm not sure I would want different functionality, but I can see the argument for changing the icon. As for the button on the right, that copies the file path to your clipboard. Perhaps I need a better notification for showing that has happened.

> The file uploader doesn't make sense to me. You click "upload file", then choose your file... then the file name gets filled in on the line for some reason? And there's no confirmation?

So I designed the file uploader to be super simple. Upload a file -> get a the url for it. That's all the functionality I thought it needed. Strictly speaking I could add a confirmation, but I don't see why you need it. The file hosting is relatively inexpensive, and if you must remove something, you can from the files interface. 

> Hidden post markers are broken, probably due to an unspecific selector on .hidden:

Yeah fixed that by changing the `hidden` class name

zwade committed:
addresses more of blupichu's concerns

zwade committed:
bug in editor upload

bluepichu commented:
Responding to your responses, hopefully in order:

> >Uploading a file from the file browser results in an inconsistent ordering: the file is added at the bottom, but on refresh, it appears at the top.

>The new element always gets put at the bottom, since its not defined as to which position it would go in, since it pulls the array from `fs.dirstat`. Now, I could sort the resulting array before building the template, but that didn't seem like a big deal to me

I'd like consistency here if possible.  Reverse order of upload time seems good to me.

> > Also, I'm not sure if the upload timestamp is part of the hash, but sometimes I can upload duplicates and sometimes I can't.

> It should get renamed if you try to upload a dupe. When is it not working for you?

Uploading directly from the file browser, I *never* got a renamed file.  I ended up with many files of (what appeared to be) the same name.  Can't speak for the editor uploader, though.

> > Also, the hash is now missing from the filename, which is weird.

> I explicitly removed the hash to make it easier to see which file is which. I prefer it that way, but I can add the hash if you want.

Now knowing that the right button copies the serving URL, this isn't necessary.

> > The left button doesn't do what I'd expect (i.e. download the file without showing it to me), and the right button doesn't seem to do anything at all.

> The left button serves the file with its proper mime type. The reason for this is so you can serve it directly. I'm not sure I would want different functionality, but I can see the argument for changing the icon. As for the button on the right, that copies the file path to your clipboard. Perhaps I need a better notification for showing that has happened.

Definitely change the left icon, and I'd also add "download" as a separate option, but I could take or leave that.

You should be able to take the CSS verbatim from the `<dttw-code>` header, which displays a "copied" animation.

> > The file uploader doesn't make sense to me. You click "upload file", then choose your file... then the file name gets filled in on the line for some reason? And there's no confirmation?

> So I designed the file uploader to be super simple. Upload a file -> get a the url for it. That's all the functionality I thought it needed. Strictly speaking I could add a confirmation, but I don't see why you need it. The file hosting is relatively inexpensive, and if you must remove something, you can from the files interface.

For a while, I didn't get why I was seeing the URL at all.  Playing with it for a while, it's more obvious now.  It makes sense, but I will say that I dislike the "disabled" color of the result line.  It also looks like it's something the user is supposed to fill in.  Maybe hide it until the user uploads a file?

Also, if possible, it'd be neat to have a per-post listing of files that were uploaded via the editor on that specific post for easy access (similar to how WordPress shows files, if I recall correctly), but that's probably not a huge concern right now.  Also neat would be a custom context menu on the editor with an "insert image" option, but again, probably outside the scope of what you're doing here.

zwade commented:
Counter-counter-counter responses? Idek anymore

> > > Uploading a file from the file browser results in an inconsistent ordering: the file is added at the bottom, but on refresh, it appears at the top.

> > The new element always gets put at the bottom, since its not defined as to which position it would go in, since it pulls the array from fs.dirstat. Now, I could sort the resulting array before building the template, but that didn't seem like a big deal to me

> I'd like consistency here if possible. Reverse order of upload time seems good to me.

Originally I thought I couldn't do this because I'm not storing metadata, but I can pull that from the filesystem. 

> Uploading directly from the file browser, I never got a renamed file. I ended up with many files of (what appeared to be) the same name. Can't speak for the editor uploader, though.

That's what the bit at the beginning is for. It's in there so you can have multiple files with the same name and not have it bork. The downside (as a result of how `multer` works) is that if you upload the same file multiple times, you get dupes. If I manually compute the hash, then  I can eliminate the dupes, but I'm not sure if this is what you're asking for

> Definitely change the left icon, and I'd also add "download" as a separate option, but I could take or leave that.

I'll pass on download for right now, since I don't think it will be super useful . If we find through use that it would be, then I'll go ahead and add that. 

> You should be able to take the CSS verbatim from the <dttw-code> header, which displays a "copied" animation.

❤️ 

> For a while, I didn't get why I was seeing the URL at all. Playing with it for a while, it's more obvious now. It makes sense, but I will say that I dislike the "disabled" color of the result line. It also looks like it's something the user is supposed to fill in. Maybe hide it until the user uploads a file?

hmm, ok. I can do that

> Also, if possible, it'd be neat to have a per-post listing of files that were uploaded via the editor on that specific post for easy access (similar to how WordPress shows files, if I recall correctly), but that's probably not a huge concern right now. Also neat would be a custom context menu on the editor with an "insert image" option, but again, probably outside the scope of what you're doing here.

I was considering that, but again, currently doing this without storing metadata.

zwade committed:
more style and ui fixes
